File: src/load_balancer/load_balancer.cxx
/**
 * GauXC Copyright (c) 2020-2024, The Regents of the University of California,
 * through Lawrence Berkeley National Laboratory (subject to receipt of
 * any required approvals from the U.S. Dept. of Energy). All rights reserved.
 *
 * See LICENSE.txt for details
 */
#include "load_balancer_impl.hpp"

namespace GauXC {

LoadBalancer::LoadBalancer( std::unique_ptr<pimpl_type>&& pimpl ): 
  pimpl_( std::move(pimpl) ) { }

LoadBalancer::LoadBalancer() : LoadBalancer( nullptr ) { }

LoadBalancer::LoadBalancer( const LoadBalancer& other ) :
  LoadBalancer(other.pimpl_->clone()){ }

LoadBalancer::LoadBalancer( LoadBalancer&& ) noexcept = default;
              
LoadBalancer::~LoadBalancer() noexcept = default;


const std::vector<XCTask>& LoadBalancer::get_tasks() const {
  if( not pimpl_ ) GAUXC_PIMPL_NOT_INITIALIZED();
  return pimpl_->get_tasks();
}
std::vector<XCTask>& LoadBalancer::get_tasks() {
  if( not pimpl_ ) GAUXC_PIMPL_NOT_INITIALIZED();
  return pimpl_->get_tasks();
}

void LoadBalancer::rebalance_weights() {
  if( not pimpl_ ) GAUXC_PIMPL_NOT_INITIALIZED();
  pimpl_->rebalance_weights();
}

void LoadBalancer::rebalance_exc_vxc() {
  if( not pimpl_ ) GAUXC_PIMPL_NOT_INITIALIZED();
  pimpl_->rebalance_exc_vxc();
}

void LoadBalancer::rebalance_exx() {
  if( not pimpl_ ) GAUXC_PIMPL_NOT_INITIALIZED();
  pimpl_->rebalance_exx();
}

const util::Timer& LoadBalancer::get_timings() const {
  if( not pimpl_ ) GAUXC_PIMPL_NOT_INITIALIZED();
  return pimpl_->get_timings();
}

size_t LoadBalancer::total_npts() const {
  if( not pimpl_ ) GAUXC_PIMPL_NOT_INITIALIZED();
  return pimpl_->total_npts();
}
size_t LoadBalancer::max_npts() const {
  if( not pimpl_ ) GAUXC_PIMPL_NOT_INITIALIZED();
  return pimpl_->max_npts();
}
size_t LoadBalancer::max_nbe() const {
  if( not pimpl_ ) GAUXC_PIMPL_NOT_INITIALIZED();
  return pimpl_->max_nbe();
}
size_t LoadBalancer::max_npts_x_nbe() const {
  if( not pimpl_ ) GAUXC_PIMPL_NOT_INITIALIZED();
  return pimpl_->max_npts_x_nbe();
}



const Molecule& LoadBalancer::molecule() const {
  if( not pimpl_ ) GAUXC_PIMPL_NOT_INITIALIZED();
  return pimpl_->molecule();
}
const MolMeta& LoadBalancer::molmeta() const {
  if( not pimpl_ ) GAUXC_PIMPL_NOT_INITIALIZED();
  return pimpl_->molmeta();
}

const LoadBalancer::basis_type& LoadBalancer::basis() const {
  if( not pimpl_ ) GAUXC_PIMPL_NOT_INITIALIZED();
  return pimpl_->basis();
}
const LoadBalancer::basis_map_type& LoadBalancer::basis_map() const {
  if( not pimpl_ ) GAUXC_PIMPL_NOT_INITIALIZED();
  return pimpl_->basis_map();
}
const LoadBalancer::shell_pair_type& LoadBalancer::shell_pairs() const {
  if( not pimpl_ ) GAUXC_PIMPL_NOT_INITIALIZED();
  return pimpl_->shell_pairs();
}

const LoadBalancer::shell_pair_type& LoadBalancer::shell_pairs() {
  if( not pimpl_ ) GAUXC_PIMPL_NOT_INITIALIZED();
  return pimpl_->shell_pairs();
}

LoadBalancerState& LoadBalancer::state() {
  if( not pimpl_ ) GAUXC_PIMPL_NOT_INITIALIZED();
  return pimpl_->state();
}

const RuntimeEnvironment& LoadBalancer::runtime() const {
  if( not pimpl_ ) GAUXC_PIMPL_NOT_INITIALIZED();
  return pimpl_->runtime();
}


bool LoadBalancer::operator==( const LoadBalancer& other ) const {
  return (&other) == this;
}

}

File: src/load_balancer.f90
! GauXC Copyright (c) 2020-2024, The Regents of the University of California,
! through Lawrence Berkeley National Laboratory (subject to receipt of
! any required approvals from the U.S. Dept. of Energy).
!
! (c) 2024-2025, Microsoft Corporation
!
! All rights reserved.
!
! See LICENSE.txt for details

!> @brief Module defining load balancing functionality for GauXC
module gauxc_load_balancer
  use iso_c_binding, only : c_ptr, c_bool, c_int, c_char, c_null_char, c_null_ptr
  use gauxc_status, only : gauxc_status_type
  use gauxc_types, only : gauxc_header_type, gauxc_type_load_balancer, &
    & gauxc_type_load_balancer_factory
  use gauxc_runtime_environment, only : gauxc_runtime_environment_type
  use gauxc_molecule, only : gauxc_molecule_type
  use gauxc_basisset, only : gauxc_basisset_type
  use gauxc_molgrid, only : gauxc_molgrid_type
  implicit none
  private

  public :: &
    & gauxc_load_balancer_factory_new, &
    & gauxc_load_balancer_factory_delete, &
    & gauxc_load_balancer_factory_get_instance, &
    & gauxc_load_balancer_factory_get_shared_instance, &
    & gauxc_load_balancer_delete

  public :: &
    & gauxc_delete, &
    & gauxc_get_instance, &
    & gauxc_get_shared_instance

  type, bind(c), public :: gauxc_load_balancer_type
    type(gauxc_header_type) :: hdr = gauxc_header_type(gauxc_type_load_balancer)
    type(c_ptr) :: ptr = c_null_ptr
    logical(c_bool) :: owned = .true.
  end type gauxc_load_balancer_type

  type, bind(c), public :: gauxc_load_balancer_factory_type
    type(gauxc_header_type) :: hdr = gauxc_header_type(gauxc_type_load_balancer_factory)
    type(c_ptr) :: ptr = c_null_ptr
  end type gauxc_load_balancer_factory_type

  interface
    !> @brief Create a new LoadBalancerFactory instance.
    function gauxc_load_balancer_factory_new_c(status, ex, kernel_name) result(lbf) &
      & bind(c, name="gauxc_load_balancer_factory_new")
      import :: gauxc_status_type, gauxc_load_balancer_factory_type, c_int, c_char
      implicit none
      !> @param status Status object to capture any errors.
      type(gauxc_status_type), intent(inout) :: status
      !> @param ex Execution space.
      integer(c_int), value :: ex
      !> @param kernel_name Name of the load balancing kernel to use.
      character(kind=c_char), intent(in) :: kernel_name(*)
      !> @return Handle to the created LoadBalancerFactory.
      type(gauxc_load_balancer_factory_type) :: lbf
    end function gauxc_load_balancer_factory_new_c
  end interface

  interface gauxc_delete
    !> @brief Delete a LoadBalancerFactory instance.
    subroutine gauxc_load_balancer_factory_delete(status, factory) bind(c)
      import :: gauxc_status_type, gauxc_load_balancer_factory_type
      implicit none
      !> @param status Status object to capture any errors.
      type(gauxc_status_type), intent(inout) :: status
      !> @param factory Handle to the LoadBalancerFactory to delete.
      type(gauxc_load_balancer_factory_type), intent(inout) :: factory
    end subroutine gauxc_load_balancer_factory_delete
  end interface gauxc_delete

  interface gauxc_get_instance
    !> @brief Create a new LoadBalancer instance from a LoadBalancerFactory.
    function gauxc_load_balancer_factory_get_instance( &
      status, factory, env, mol, mg, basis) result(lb) bind(c)
      import :: gauxc_status_type, gauxc_load_balancer_factory_type, &
        & gauxc_load_balancer_type, gauxc_molecule_type, gauxc_molgrid_type, &
        & gauxc_basisset_type, gauxc_runtime_environment_type, c_ptr
      implicit none
      !> @param status Status object to capture any errors.
      type(gauxc_status_type), intent(inout) :: status
      !> @param factory Handle to the LoadBalancerFactory.
      type(gauxc_load_balancer_factory_type), value :: factory
      !> @param env Handle to the RuntimeEnvironment.
      type(gauxc_runtime_environment_type), value :: env
      !> @param mol Handle to the Molecule.
      type(gauxc_molecule_type), value :: mol
      !> @param mg Handle to the MolGrid.
      type(gauxc_molgrid_type), value :: mg
      !> @param basis Handle to the BasisSet.
      type(gauxc_basisset_type), value :: basis
      !> @return Handle to the created LoadBalancer.
      type(gauxc_load_balancer_type) :: lb
    end function gauxc_load_balancer_factory_get_instance
  end interface gauxc_get_instance

  interface gauxc_get_shared_instance
    !> @brief Create a shared LoadBalancer instance from a LoadBalancerFactory.
    function gauxc_load_balancer_factory_get_shared_instance( &
      status, factory, env, mol, mg, basis) result(lb) bind(c)
      import :: gauxc_status_type, gauxc_load_balancer_factory_type, &
        & gauxc_load_balancer_type, gauxc_molecule_type, gauxc_molgrid_type, &
        & gauxc_basisset_type, gauxc_runtime_environment_type, c_ptr
      implicit none
      !> @param status Status object to capture any errors.
      type(gauxc_status_type), intent(inout) :: status
      !> @param factory Handle to the LoadBalancerFactory.
      type(gauxc_load_balancer_factory_type), value :: factory
      !> @param env Handle to the RuntimeEnvironment.
      type(gauxc_runtime_environment_type), value :: env
      !> @param mol Handle to the Molecule.
      type(gauxc_molecule_type), value :: mol
      !> @param mg Handle to the MolGrid.
      type(gauxc_molgrid_type), value :: mg
      !> @param basis Handle to the BasisSet.
      type(gauxc_basisset_type), value :: basis
      !> @return Handle to the created LoadBalancer.
      type(gauxc_load_balancer_type) :: lb
    end function gauxc_load_balancer_factory_get_shared_instance
  end interface gauxc_get_shared_instance

  interface gauxc_delete
    !> @brief Delete a LoadBalancer instance.
    subroutine gauxc_load_balancer_delete(status, lb) bind(c)
      import :: gauxc_status_type, gauxc_load_balancer_type
      implicit none
      !> @param status Status object to capture any errors.
      type(gauxc_status_type), intent(inout) :: status
      !> @param lb Handle to the LoadBalancer to delete.
      type(gauxc_load_balancer_type), intent(inout) :: lb
    end subroutine gauxc_load_balancer_delete
  end interface gauxc_delete

contains

  !> @brief Create a new LoadBalancerFactory instance.
  function gauxc_load_balancer_factory_new(status, ex, kernel_name) result(lbf)
    !> @param status Status object to capture any errors.
    type(gauxc_status_type), intent(inout) :: status
    !> @param ex Execution space.
    integer(c_int), value :: ex
    !> @param kernel_name Name of the load balancing kernel to use.
    character(kind=c_char, len=*), intent(in) :: kernel_name
    !> @return Handle to the created LoadBalancerFactory.
    type(gauxc_load_balancer_factory_type) :: lbf

    character(kind=c_char), allocatable :: c_kernel_name(:)

    c_kernel_name = transfer(kernel_name // c_null_char, &
      & [character(kind=c_char) ::], len(kernel_name) + 1)

    lbf = gauxc_load_balancer_factory_new_c(status, ex, c_kernel_name)
  end function gauxc_load_balancer_factory_new
end module gauxc_load_balancer
